### PR TITLE
Fix pauli symmetries

### PR DESCRIPTION
**Pull request by RickSimon79221**
## 📝 Description
In previous big merge, Pauli symmetry code was not updated to new syntax. That is done here. Also fixes bug in the generation of random pauli-symmtric hamiltonians with redundant qubits. Also adds documentation and tests for pauli symmetry code 

## ✅ Checklist before requesting a review

- [x] I have made corresponding changes to the documentation.
- [x] The code follows the defined style guide.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] The code passes CI.

### CODE REVIEW

File: src/sympleq/core/symmetries/pauli.py [hunk 1]
```diff
@@ -1,15 +1,26 @@
 import numpy as np
 import itertools
 from sympleq.core.paulis import PauliSum
-from sympleq.core.circuits import Circuit, Hadamard as H, SUM as CX, PHASE as S
+from sympleq.core.circuits import Circuit, HADAMARD as H, CX, PHASE as S
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py]: I think we should import `GATES` and use the gates as `GATES.H` rather than instatiating new singletons (i give an example later)

File: src/sympleq/core/symmetries/pauli.py [hunk 2]
```diff
@@ -40,53 +85,133 @@ def number_of_S(x_exp, z_exp, d):
...
         if pauli_sum.x_exp[pauli_index, i]:
             number_of_sum_x = number_of_SUM_X(pauli_sum.x_exp[pauli_index, qudit],
                                               pauli_sum.x_exp[pauli_index, i],
                                               pauli_sum.dimensions[i])
-            list_of_gates += [CX(qudit, i, pauli_sum.dimensions[qudit])] * number_of_sum_x
-    C.add_gate(list_of_gates)
-    for g in list_of_gates:
-        pauli_sum = g.act(pauli_sum)
+            for _ in range(number_of_sum_x):
+                C.add_gate(CX(), qudit, i)
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py]:
```suggestion
                C.add_gate(GATES.CX, qudit, i)
```

File: src/sympleq/core/symmetries/pauli.py [hunk 3]
```diff
@@ -40,53 +85,133 @@ def number_of_S(x_exp, z_exp, d):
...
+                pauli_sum = CX().act(pauli_sum, (qudit, i))
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py]:
```suggestion
                pauli_sum = GATES.CX.act(pauli_sum, (qudit, i))
```

File: src/sympleq/core/symmetries/pauli.py [hunk 4]
```diff
@@ -40,53 +85,133 @@ def number_of_S(x_exp, z_exp, d):
...
+    C : Circuit
+        The circuit after the gates have been added to cancel out the Z part of the Pauli operator.
+    """
+    C.add_gate(H(), qudit)
+    pauli_sum = H().act(pauli_sum, qudit)
     for i in range(qudit + 1, q_max):
         if pauli_sum.z_exp[pauli_index, i]:
             number_of_sum_z = number_of_SUM_Z(pauli_sum.z_exp[pauli_index, i],
-                                              pauli_sum.x_exp[pauli_index, qudit],
+                                              pauli_sum.z_exp[pauli_index, qudit],
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py:152]: this seems a substantial change, Im not familiar with this code, but are u sure we need to change x to z here?

File: src/sympleq/core/symmetries/pauli.py [hunk 5]
```diff
@@ -40,53 +85,133 @@ def number_of_S(x_exp, z_exp, d):
...
+    number_of_phase = number_of_S(pauli_sum.x_exp[pauli_index, qudit], pauli_sum.z_exp[pauli_index, qudit],
+                                  pauli_sum.dimensions[qudit])
+    for _ in range(number_of_phase):
+        C.add_gate(S(), qudit)
+        pauli_sum = S().act(pauli_sum, qudit)
     return pauli_sum, C
 
 
-def cancel_pauli(P, current_qudit, pauli_index, circuit, n_q_max):
+def cancel_pauli(P: PauliSum, current_qudit: int, pauli_index: int, circuit: Circuit, n_q_max: int):
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py]: missing return types

File: src/sympleq/core/symmetries/pauli.py [hunk 6]
```diff
@@ -138,9 +303,8 @@ def symplectic_reduction_iter_qudit_(P, C, pivots, current_qudit):
     # does the current qudit have any X or Z components?
     if any(P.x_exp[:, current_qudit]) or any(P.z_exp[:, current_qudit]):
         if not any(P.x_exp[:, current_qudit]):  # If it is z we need to add a Hadamard gate to make it an X
-            g = H(current_qudit, P.dimensions[current_qudit])
-            C.add_gate(g)
-            P = g.act(P)
+            C.add_gate(H(), current_qudit)
```
Comment (ricott1) [inline src/sympleq/core/symmetries/pauli.py]: We could also add a new API as discussed (in a different PR), to do smtg like `C.H(current_qudit)`, what you think?

File: src/sympleq/core/symmetries/pauli.py (hunk 4 above)
Comment (ldellant) [reply src/sympleq/core/symmetries/pauli.py:152]: Good catch, tbh I am not sure whether before was wrong or not. I think we never tested this function cause we are always diagonalizing in the Z basis. _If it is fast_ we could add a test where we diagonalize in the X basis and we check that all Z part is zero...

Comment (ldellant) [reply src/sympleq/core/symmetries/pauli.py:152]: Actually not sure about this, looking at `cancel_pauli`... Just ignore me :)

File: src/sympleq/core/symmetries/pauli.py (hunk 5 above)
Comment (ldellant) [reply src/sympleq/core/symmetries/pauli.py]: Ignore this comment, but @ricott1 if you could explain why it is missing I would be happy :) 

Seems it returns a PauliSum and a Circuit, which are initialized?

File: src/sympleq/core/symmetries/pauli.py [hunk 7]
```diff
@@ -152,31 +316,55 @@ def symplectic_reduction_iter_qudit_(P, C, pivots, current_qudit):
...
 
-        g = H(current_qudit, P.dimensions[current_qudit])
-        C.add_gate(g)
-        P = g.act(P)
+        C.add_gate(H(), current_qudit)
+        P = H().act(P, current_qudit)
     return C, pivots
 
 
 def symplectic_pauli_reduction(hamiltonian: PauliSum) -> Circuit:
```
Comment (ldellant) [inline src/sympleq/core/symmetries/pauli.py]: Can we eliminate this function? Seems redundant, as we can call `symplectic_reduction_qudit` and select C?

File: src/sympleq/core/symmetries/pauli.py [hunk 8]
```diff
@@ -152,31 +316,55 @@ def symplectic_reduction_iter_qudit_(P, C, pivots, current_qudit):
...
+    Returns
+    -------
+    Circuit
+        The Circuit that implements the symplectic reduction algorithm.
+    """
     C, pivots = symplectic_reduction_qudit(hamiltonian)
     return C
 
 
 def pauli_reduce(hamiltonian: PauliSum) -> tuple[PauliSum, list[PauliSum], Circuit, list]:
```
Comment (ldellant) [inline src/sympleq/core/symmetries/pauli.py:330]: I almost wrote you to eliminate this function cause you have written the class for conditional Hamiltonian, that is just much better... I think however it is worth keeping for testing purposes :)

File: src/sympleq/models/random_hamiltonian.py [hunk 9]
```diff
@@ -181,8 +180,9 @@ def random_pauli_symmetry_hamiltonian(n_qudits: int, n_paulis: int, n_redundant=
 
     P = PauliSum.from_string(pauli_strings, dimensions=[2] * n_qudits, weights=weights, phases=phases)
 
-    g = Gate.from_random(n_qudits, 2)
-    P = g.act(P, 2)
+    if shuffle:
+        g = Gate.from_random(n_qudits, 2)
```
Comment (ldellant) [inline src/sympleq/models/random_hamiltonian.py:184]: let's go a bit heavier with the gates... `10 * n_qudits ** 2`

File: tests/core_tests/symmetries_tests/symmetries_test.py [hunk 10]
```diff
@@ -159,3 +160,22 @@ def test_random_arbitrary_symmetry(self):
                 assert np.all(H_s.weights == H_out.weights)
 
                 assert c.act(H, all_qudit_indices).to_standard_form() == H.to_standard_form()
+
+    def test_random_pauli_symmetry(self):
+        n_tests = 10
+        n_qudits = 10
```
Comment (ldellant) [inline tests/core_tests/symmetries_tests/symmetries_test.py]: `n_ qudits = rng.integers(5, 25)`

File: tests/core_tests/symmetries_tests/symmetries_test.py [hunk 11]
```diff
@@ -159,3 +160,22 @@ def test_random_arbitrary_symmetry(self):
...
+        n_paulis = 50
```
Comment (ldellant) [inline tests/core_tests/symmetries_tests/symmetries_test.py]: `n_ paulis = rng.integers(5 * n_qudits, 10 * n_qudits ** 2)`

**Review by ldellant — APPROVED** [review]

File: src/sympleq/core/symmetries/pauli.py (hunk 5 above)
Comment (ricott1) [reply src/sympleq/core/symmetries/pauli.py]: I meant adding the explicit return type as `-> type` at the end of the line

File: tests/core_tests/symmetries_tests/symmetries_test.py (hunk 10 above)
Comment (ricott1) [reply tests/core_tests/symmetries_tests/symmetries_test.py]: In general, I don't like using randomness in unit tests, and I think we should rather cover as much as possible and then have separate fuzzing tests. but no strong opinion

File: src/sympleq/core/symmetries/pauli.py (hunk 4 above)
Comment (RickSimon79221) [reply src/sympleq/core/symmetries/pauli.py:152]: Yes, before, all the paulis where first put into a list and only executet after being collected. This way of doing it seemed to be harder/more annoying to do with the new inplementation of the gates and circuits, so instead I apply the gates now directly, as they are collected. Specifically here, the first gate is a Hadamard that takes the X in the qudit used to eliminate everything else and turns it into a Z. So, if we don't apply the Hadamard immediatly, the X-exponent is the important quantity, if we do, then it becomes the Z-exponent

File: src/sympleq/core/symmetries/pauli.py (hunk 6 above)
Comment (RickSimon79221) [reply src/sympleq/core/symmetries/pauli.py]: Yeah, could be useful

File: src/sympleq/models/random_hamiltonian.py (hunk 9 above)
Comment (RickSimon79221) [reply src/sympleq/models/random_hamiltonian.py:184]: Its not a random circuit, its a random gate of given size. It is already as random as it can get (probably).

**Review by ricott1 — APPROVED** [review]
